test/Feature/RpcSpec.hs

```diff
@@ -96,22 +96,113 @@ spec actualPgVersion =
+          "hint":"If a new function was created in the database with this name and arguments, try reloading the schema cache.",
+          "message":"Couldn't find the test.add_them(a, b, smthelse) function" } |]
+        { matchStatus  = 404
+        , matchHeaders = [matchContentTypeJson]
+        }
+
+      it "should fail with 404 when no json arg is found with prefer single object" $
+        request methodPost "/rpc/sayhello"
+          [("Prefer","params=single-object")]
+          [json|{}|]
+        `shouldRespondWith`
+          [json| {
+            "hint":"If a new function was created in the database with this name and arguments, try reloading the schema cache.",
+            "message":"Couldn't find the test.sayhello() function" } |]
         { matchStatus  = 404
         , matchHeaders = [matchContentTypeJson]
         }
 
+      it "should fail with 404 for overloaded functions with unknown args" $ do
+        get "/rpc/overloaded?wrong_arg=value" `shouldRespondWith`
+          [json| {
+            "hint":"If a new function was created in the database with this name and arguments, try reloading the schema cache.",
+            "message":"Couldn't find the test.overloaded(wrong_arg) function" } |]
+          { matchStatus  = 404
+          , matchHeaders = [matchContentTypeJson]
+          }
+        get "/rpc/overloaded?a=1&b=2&wrong_arg=value" `shouldRespondWith`
+          [json| {
+            "hint":"If a new function was created in the database with this name and arguments, try reloading the schema cache.",
+            "message":"Couldn't find the test.overloaded(a, b, wrong_arg) function" } |]
+          { matchStatus  = 404
+          , matchHeaders = [matchContentTypeJson]
+          }
+
+      it "should fail with 404 for unkown explicit argument type casts on overloaded functions" $
+        get "/rpc/overloaded_same_args?arg::numeric=123" `shouldRespondWith`
+        [json| {
+          "hint":"If a new function was created in the database with this name and arguments, try reloading the schema cache.",
+          "message":"Couldn't find the test.overloaded_same_args(arg::numeric) function" } |]
+        { matchStatus  = 404
+        , matchHeaders = [matchContentTypeJson]
+        }
+
+    context "ambiguous overloaded functions with same arguments but different types" $ do
+      it "should fail with 300 Multiple Choises without explicit argument type casts" $
+        get "/rpc/overloaded_same_args?arg=value" `shouldRespondWith`
+          [json| {
+            "hint":"Explicit argument type casts are needed. Possible endpoints are: /rpc/overloaded_same_args?arg::integer=value , /rpc/overloaded_same_args?arg::xml=value , /rpc/overloaded_same_args?arg::text=value&num::integer=value",
+            "message":"Could not choose the best candidate function between: test.overloaded_same_args(arg => integer), test.overloaded_same_args(arg => xml), test.overloaded_same_args(arg => text, num => integer)" } |]
```

Comment:
Typo in `300 Multiple Choises` - should be `Choices`.

Btw, really nice error message! This is much better than the ambiguous error for multiple relationships.